nirc_ehr/resources/queries/study/assignment.js

```diff
@@ -2,61 +2,95 @@ require("ehr/triggers").initScript(this);
+var batchLatestEnddate;
+var count = 0;
 
-function onInit(event, helper){
+function getLastAssignment(id){
     LABKEY.Query.selectRows({
-        schemaName: 'ehr',
-        queryName: 'project',
-        columns: 'project,name',
-        success: function(results){
-            if (results.rows.length){
+        schemaName: 'study',
+        queryName: 'assignment',
+        columns: 'Id,enddate',
+        filterArray: [LABKEY.Filter.create('Id', id)],
+        success: function (results) {
+            if (results.rows.length) {
                 for (var i = 0; i < results.rows.length; i++) {
                     let rec = results.rows[i];
-                    projectData[rec.name] = rec.project;
+                    if (!batchLatestEnddate) {
+                        batchLatestEnddate = rec.enddate;
+                    }
+                    else {
+                        var oldDate = new Date(batchLatestEnddate);
+                        var newDate = new Date(rec.enddate);
+                        if (newDate > oldDate) {
+                            batchLatestEnddate = rec.enddate;
+                        }
+                    }
                 }
             }
         },
         scope: this
     });
 }
 
-EHR.Server.TriggerManager.registerHandlerForQuery(EHR.Server.TriggerManager.Events.BEFORE_INSERT, 'study', 'assignment', function (helper, scriptErrors, row, oldRow) {
-
-    // use enddate of previous assignment otherwise animal birth/arrival will be in row.date
-    if (prevAnimalId === row.Id) {
-        row.date = prevDate;
+function onInit(event, helper){
+    if (helper.isETL()) {
+        LABKEY.Query.selectRows({
+            schemaName: 'ehr',
+            queryName: 'project',
+            columns: 'project,name',
+            success: function (results) {
+                if (results.rows.length) {
+                    for (var i = 0; i < results.rows.length; i++) {
+                        let rec = results.rows[i];
+                        projectData[rec.name] = rec.project;
+                    }
+                }
+            },
+            scope: this
+        });
     }
+}
 
-    if (row.changeDateText) {
-        if (row.changeDateText.indexOf("ACQ:") === 0) {
-            row.endDate = new Date(Date.parse(row.changeDateText.split(":")[1].trim()));
+EHR.Server.TriggerManager.registerHandlerForQuery(EHR.Server.TriggerManager.Events.BEFORE_INSERT, 'study', 'assignment', function (helper, scriptErrors, row, oldRow) {
+
+    if (helper.isETL()) {
+        if (prevAnimalId === row.Id) {
+            row.date = prevDate;
         }
-        else {
-            row.endDate = new Date(Date.parse(row.changeDateText));
+        else if (count === 0) {
+            // This handles batch boundary row for full truncate ETL, which is the only ETL setup for this currently.
+            // Gets previous enddate from db for first row
+            getLastAssignment(row.Id);
+            if (batchLatestEnddate) {
```

Comment:
Is there a reason to have batchLatestEnddate as a top-level variable instead of returning it from getLastAssignment()?

Reply:
I refactored this to return instead of  the top level variable